storage/libradb/src/pruner/mod.rs

```diff
@@ -0,0 +1,265 @@
+// Copyright (c) The Libra Core Contributors
+// SPDX-License-Identifier: Apache-2.0
+
+//! This module provides `Pruner` which manages a thread pruning old data in the background and is
+//! meant to be triggered by other threads as they commit new data to the DB.
+
+use crate::{
+    schema::{
+        retired_state_record::RetiredStateRecordSchema, state_merkle_node::StateMerkleNodeSchema,
+    },
+    OP_COUNTER,
+};
+use failure::prelude::*;
+use logger::prelude::*;
+use schemadb::{ReadOptions, SchemaBatch, DB};
+use sparse_merkle::RetiredRecordType;
+use std::{
+    sync::{
+        mpsc::{channel, Receiver, Sender},
+        Arc, Mutex,
+    },
+    thread::JoinHandle,
+};
+use types::transaction::Version;
+
+use failure::_core::sync::atomic::Ordering;
+use std::sync::atomic::AtomicU64;
+#[cfg(test)]
+use std::thread::sleep;
+#[cfg(test)]
+use std::time::{Duration, Instant};
+
+/// The `Pruner` is meant to be part of a `LibraDB` instance and runs in the background to prune old
+/// data.
+///
+/// It creates a worker thread on construction and joins it on destruction. When destructed, it
+/// quits the worker thread eagerly without waiting for all pending work to be done.
+pub(crate) struct Pruner {
+    /// Other than the latest version, how many historical versions to keep being readable. For
+    /// example, this being 0 means keep only the latest version.
+    num_historical_versions_to_keep: u64,
+    /// The worker thread handle, created upon Pruner instance construction and joined upon its
+    /// destruction. It only becomes `None` after joined in `drop()`.
+    worker_thread: Option<JoinHandle<()>>,
+    /// The sender side of the channel talking to the worker thread.
+    command_sender: Mutex<Sender<Command>>,
+    /// (For tests) A way for the worker thread to inform the `Pruner` the pruning progress. If it
+    /// sets this atomic value to `V`, all versions before `V` can no longer be accessed.
+    #[allow(dead_code)]
```

Comment:
`#[cfg(test)]`?

Reply:
It's not easy to do so since I initialize and pass it to the worker in the constructor.